app/models/product.rb

```diff
@@ -21,11 +13,13 @@ def validate_each(record, attribute, value)
     end
   end
 end
+
 class Product < ApplicationRecord
 
-    has_many :line_items
-    has_many  :orders, through: :line_items
-    before_destroy :ensure_not_referenced_by_any_line_item
+    has_many :line_items, dependent: :restrict_with_error
```

Comment:
@avneetsng Check all dependent options values.

Reply:
1.  **`destroy_async`** behaviour is inconsistent and the **`destroy_async`** option should not be used when the association is backed by foreign key constraints in the database, it will throw
`FOREIGN KEY constraint failed (ActiveRecord::InvalidForeignKey)`
2. `dependent: :nullify` → If we try to add this and the foreign key has set `null: false`
then destroying the ActiveRecord will give 
    - `SQLite3::ConstraintException: NOT NULL constraint failed: books.author_id (ActiveRecord::NotNullViolation)`
    
3. `dependent: :restrict_with_exception` → throws this exception `(ActiveRecord::DeleteRestrictionError)` when you try to delete the record
4. `dependent: :restrict_with_error` → It won't raise any exception just throws an error that will make the transaction rollback